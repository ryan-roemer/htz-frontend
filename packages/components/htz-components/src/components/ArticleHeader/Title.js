import React from 'react';
import PropTypes from 'prop-types';
import { createComponent, } from 'react-fela';
import { parseStyleProps, parseTypographyProp, } from '@haaretz/htz-css-tools';
import { responsivePropBaseType, } from '../../propTypes/responsivePropBaseType';
import { stylesPropType, } from '../../propTypes/stylesPropType';

const titlePropTypes = {
  /**
   * Sets heading element to be _block_ or _inline_.(default is inline)
   */
  isBlock: PropTypes.bool,
  /**
   * Font-size configuration for the heading.<br/>
   * Can be an int, defined at: [theme typography](https://haaretz.github.io/htz-frontend/docs/htz-theme/#typographic-scale)
   * Or an array of responsive-prop-object.
   * ```js
   *  { from, until, value, }
   * ```
   */
  fontSize: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.arrayOf(
      PropTypes.shape({
        ...responsivePropBaseType,
        value: PropTypes.number,
      })
    ),
  ]).isRequired,
  /**
   * Heading level. The h`${level}` tag to create.
   */
  level: PropTypes.oneOf([ 1, 2, 3, 4, 5, 6, ]).isRequired,
  /**
   * Text of title
   */
  text: PropTypes.string.isRequired,
  /**
   * A special property holding miscellaneous CSS values that
   * trumps all default values. Processed by
   * [`parseStyleProps`](https://Haaretz.github.io/htz-frontend/htz-css-tools#parsestyleprops)
   */
  miscStyles: stylesPropType,
};

const titleDefaultProps = {
  isBlock: false,
  miscStyles: null,
};

/**
 * See Title.propTypes of details
 * @param {object} props
 * @param {boolean} props.isBlock
 *   Sets heading element to be _block_ or _inline_.(default is inline)
 * @param {number|Array} props.fontSize
 *   Font-size configuration for the heading
 * @param {object} props.miscStyles
 *   A special property holding miscellaneous CSS values
 * @param {object} props.theme
 *   theme :-)
 */
const styleTitle = ({ isBlock, fontSize, miscStyles, theme, }) => ({
  display: isBlock ? 'block' : 'inline',
  extend: [
    parseTypographyProp(fontSize, theme.type),
    ...(miscStyles ? parseStyleProps(miscStyles) : []),
  ],
});

// eslint-disable-next-line react/prop-types
function TitleComponent({ level, className, text, }) {
  const HeadingTag = `h${level}`;
  return <HeadingTag className={className}>{text}</HeadingTag>;
}

const TitleStyled = createComponent(styleTitle, TitleComponent, props =>
  Object.keys(props)
);

function Title({ isBlock, fontSize, level, text, miscStyles, }) {
  return (
    <TitleStyled
      isBlock={isBlock}
      fontSize={fontSize}
      level={level}
      text={text}
      miscStyles={miscStyles}
    />
  );
}

Title.propTypes = titlePropTypes;
Title.defaultProps = titleDefaultProps;
export default Title;
