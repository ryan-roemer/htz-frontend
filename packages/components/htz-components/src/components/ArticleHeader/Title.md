### Title Component

A generic title component for use within teasers and articles.
Creates an HTML _`H(n)`_ tag. (H1, H2, H3 ...)

```jsx
<Title
  isBlock={false}
  fontSize={[
    { until: 'm', value: 3, },
    { from: 'm', until: 'l', value: 4, },
    { from: 'l', value: 5, },
  ]}
  level={1}
  text="ישראל שומרת על ריסון בעזה ומאמצת גישה תקיפה בגבול הצפון"
/>
```
