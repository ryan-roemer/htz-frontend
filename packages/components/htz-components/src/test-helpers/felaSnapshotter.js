import React from 'react';
import createRenderer from '../components/StyleProvider/createRenderer';
import StyleProvider from '../components/StyleProvider/StyleProvider';
import createFelaSnapshotter from './createFelaSnapshotter';

const testRenderer = createRenderer({ isRtl: true, });
const felaSnapshotter = createFelaSnapshotter(StyleProvider, testRenderer);

export default felaSnapshotter;
