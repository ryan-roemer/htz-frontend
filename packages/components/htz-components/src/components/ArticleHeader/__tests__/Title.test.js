import React from 'react';
import felaSnapshotter from '../../../test-helpers/felaSnapshotter';
import Title from '../Title'; // eslint-disable-line import/no-named-as-default

describe('Title component', () => {
  it('renders correctly a H1-> H6 tag, Zero font-size scale, as inline element', () => {
    for (let i = 1; i < 7; i++) {
      let snapshot = felaSnapshotter(
        <Title
          fontSize={0}
          level={i}
          text={`This is a H${i} title`}
        />
      );
      expect(snapshot).toMatchSnapshot();
    }
  });

  it('renders correctly a H1 tag as block element "display: block"', () => {
    const { component, styles, } = felaSnapshotter(
      <Title
        isBlock={true}
        fontSize={0}
        level={1}
        text="This is a block H1 title"
      />
    );
    expect(component).toMatchSnapshot();
    expect(styles).toMatchSnapshot();
  });

  it('renders correctly a H1 tag with responsive font-size', () => {
    const { component, styles, } = felaSnapshotter(
      <Title
        isBlock={true}
        fontSize={[
                { until: 'm', value: 3, },
                { from: 'm', until: 'l', value: 4, },
                { from: 'l', value: 5, },
          ]}
        level={1}
        text="This is a block H1 title"
      />
    );
    expect(component).toMatchSnapshot();
    expect(styles).toMatchSnapshot();
  });
});
