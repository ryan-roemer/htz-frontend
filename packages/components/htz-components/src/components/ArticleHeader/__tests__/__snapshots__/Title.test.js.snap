// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Title component renders correctly a H1 tag as block element "display: block" 1`] = `
<div>
  <h1
    className="htz-a htz-b htz-c htz-d htz-e"
  >
    This is a block H1 title
  </h1>
</div>
`;

exports[`Title component renders correctly a H1 tag as block element "display: block" 2`] = `".htz-a{display:block}@media (max-width: 79.9375em){.htz-b{font-size:2.6666666666666665rem}.htz-c{line-height:1.5em}}@media (min-width: 80em){.htz-d{font-size:2.5714285714285716rem}.htz-e{line-height:1.5555555555555556em}}"`;

exports[`Title component renders correctly a H1 tag with responsive font-size 1`] = `
<div>
  <h1
    className="htz-a htz-b htz-c htz-d htz-e htz-f htz-g htz-h htz-i"
  >
    This is a block H1 title
  </h1>
</div>
`;

exports[`Title component renders correctly a H1 tag with responsive font-size 2`] = `".htz-a{display:block}@media (max-width: 47.9375em){.htz-b{font-size:4rem}.htz-c{line-height:1.25em}}@media (min-width: 48em) and (max-width: 63.9375em){.htz-d{font-size:4.666666666666667rem}.htz-e{line-height:1.2857142857142858em}}@media (min-width: 64em) and (max-width: 79.9375em){.htz-f{font-size:5.333333333333333rem}.htz-g{line-height:1.125em}}@media (min-width: 80em){.htz-h{font-size:5.142857142857143rem}.htz-i{line-height:1.1666666666666667em}}"`;

exports[`Title component renders correctly a H1-> H6 tag, Zero font-size scale, as inline element 1`] = `
Object {
  "component": <div>
    <h1
      className="htz-a htz-b htz-c htz-d htz-e"
    >
      This is a H1 title
    </h1>
  </div>,
  "styles": ".htz-a{display:inline}@media (max-width: 79.9375em){.htz-b{font-size:2.6666666666666665rem}.htz-c{line-height:1.5em}}@media (min-width: 80em){.htz-d{font-size:2.5714285714285716rem}.htz-e{line-height:1.5555555555555556em}}",
}
`;

exports[`Title component renders correctly a H1-> H6 tag, Zero font-size scale, as inline element 2`] = `
Object {
  "component": <div>
    <h2
      className="htz-a htz-b htz-c htz-d htz-e"
    >
      This is a H2 title
    </h2>
  </div>,
  "styles": ".htz-a{display:inline}@media (max-width: 79.9375em){.htz-b{font-size:2.6666666666666665rem}.htz-c{line-height:1.5em}}@media (min-width: 80em){.htz-d{font-size:2.5714285714285716rem}.htz-e{line-height:1.5555555555555556em}}",
}
`;

exports[`Title component renders correctly a H1-> H6 tag, Zero font-size scale, as inline element 3`] = `
Object {
  "component": <div>
    <h3
      className="htz-a htz-b htz-c htz-d htz-e"
    >
      This is a H3 title
    </h3>
  </div>,
  "styles": ".htz-a{display:inline}@media (max-width: 79.9375em){.htz-b{font-size:2.6666666666666665rem}.htz-c{line-height:1.5em}}@media (min-width: 80em){.htz-d{font-size:2.5714285714285716rem}.htz-e{line-height:1.5555555555555556em}}",
}
`;

exports[`Title component renders correctly a H1-> H6 tag, Zero font-size scale, as inline element 4`] = `
Object {
  "component": <div>
    <h4
      className="htz-a htz-b htz-c htz-d htz-e"
    >
      This is a H4 title
    </h4>
  </div>,
  "styles": ".htz-a{display:inline}@media (max-width: 79.9375em){.htz-b{font-size:2.6666666666666665rem}.htz-c{line-height:1.5em}}@media (min-width: 80em){.htz-d{font-size:2.5714285714285716rem}.htz-e{line-height:1.5555555555555556em}}",
}
`;

exports[`Title component renders correctly a H1-> H6 tag, Zero font-size scale, as inline element 5`] = `
Object {
  "component": <div>
    <h5
      className="htz-a htz-b htz-c htz-d htz-e"
    >
      This is a H5 title
    </h5>
  </div>,
  "styles": ".htz-a{display:inline}@media (max-width: 79.9375em){.htz-b{font-size:2.6666666666666665rem}.htz-c{line-height:1.5em}}@media (min-width: 80em){.htz-d{font-size:2.5714285714285716rem}.htz-e{line-height:1.5555555555555556em}}",
}
`;

exports[`Title component renders correctly a H1-> H6 tag, Zero font-size scale, as inline element 6`] = `
Object {
  "component": <div>
    <h6
      className="htz-a htz-b htz-c htz-d htz-e"
    >
      This is a H6 title
    </h6>
  </div>,
  "styles": ".htz-a{display:inline}@media (max-width: 79.9375em){.htz-b{font-size:2.6666666666666665rem}.htz-c{line-height:1.5em}}@media (min-width: 80em){.htz-d{font-size:2.5714285714285716rem}.htz-e{line-height:1.5555555555555556em}}",
}
`;
